### PR TITLE
[Core] Add bidirectional and duplex matrix code

### PR DESCRIPTION
**Pull request by elfmimi**
## Description

Two things.

1st) Now that we have more than one keyboards that utilize duplexing of matrix by deploying both **COL2ROW** and **ROW2COL** at the same time. Namely those keyboards are:

- https://github.com/qmk/qmk_firmware/tree/master/keyboards/angel64
- https://github.com/qmk/qmk_firmware/tree/master/keyboards/thedogkeyboard

I suggest bringing the so to called COL2ROW2COL code into the core. I labeled it **BOTHWAYS** .
I've asked the author of those keyboards to test this code. and the result was perfect.

2nd) Using the fact that the technique above fully functions without ghost and with NKRO-bility, it can be proved that key-matrix with inconsistent diode direction can also be made to function, provided proper lines of code. It's worth noting the code for this is also compatible with both **COL2ROW** and **ROW2COL** at the same time. I labeled it **EITHERWAY** .

~~I suggest using **EITHERWAY** as a new default for when DIODE_DIRECTION is not defined.~~
This part is taken back.

This PR will not increase footprints of flash-rom or ram for any of the existing keyboards. It even reduces a few bytes for some split keyboards utilizing split_common.

I've also tested **BOTHWAYS** and **EITHERWAY** with my one-off experimental keyboard which has STM32F303, and it too worked great.

Minimum extent of documentation is included.
I'm planning to do extended and in-detail guide later on. after this is accepted.

- Refer to https://github.com/qmk/qmk_firmware/pull/8115 for the first commit of the two. This is to stop Travis from complaining.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by elfmimi** [comment]
Schematic drawing of Angel64 can be found here.
https://github.com/kakunpc/angel/blob/master/angel64/debug/README.md

Concerning documentation the part I want to change the most is this line.

> So the fullsize matrix shown above would be possible on a Proton C or Teensy++, but not on a regular Teensy or Pro Micro

https://docs.qmk.fm/#/hand_wire

**Fullsize keyboard is possible with a single ProMicro.**

File: quantum/config_common.h [hunk 1]
```diff
@@ -19,6 +19,8 @@
 /* diode directions */
 #define COL2ROW 0
 #define ROW2COL 1
+#define EITHERWAY 2 /* Compensation for diode direction dissonance */
+#define BOTHWAYS 3 /* Duplex-Matrix */
```
Comment (fauxpark) [inline quantum/config_common.h]: IMO we should keep the existing naming convention here, so `COL2ROW2COL` and `ROW2COL2ROW`. The documentation can explain what these mean, as they do already for `COL2ROW` and `ROW2COL`.

File: quantum/matrix.c [hunk 2]
```diff
@@ -23,14 +23,43 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-#elif (DIODE_DIRECTION == ROW2COL) || (DIODE_DIRECTION == COL2ROW)
+static const pin_t direct_pins[MATRIX_ROWS][MATRIX_COLS] = DIRECT_PINS;
+#else
+/* Let's default to EITHERWAY. This works regardless of COL2ROW or ROW2COL wiring. */
+#    if !defined(DIODE_DIRECTION)
+#        define DIODE_DIRECTION EITHERWAY
+#    endif
+
+/* Workaround for facew : MATRIX_COL_PINS has fewer number of elements than MATRIX_COLS */
+#    if defined(KEYBOARD_facew)
```
Comment (fauxpark) [inline quantum/matrix.c]: This should be fixed in the keyboard code, not in core.

File: quantum/matrix.c [hunk 3]
```diff
@@ -23,14 +23,43 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "quantum.h"
 
 #ifdef DIRECT_PINS
-static pin_t direct_pins[MATRIX_ROWS][MATRIX_COLS] = DIRECT_PINS;
-#elif (DIODE_DIRECTION == ROW2COL) || (DIODE_DIRECTION == COL2ROW)
+static const pin_t direct_pins[MATRIX_ROWS][MATRIX_COLS] = DIRECT_PINS;
+#else
+/* Let's default to EITHERWAY. This works regardless of COL2ROW or ROW2COL wiring. */
```
Comment (fauxpark) [inline quantum/matrix.c]: I would prefer if there was no default direction. See https://github.com/qmk/qmk_firmware/pull/7915

Comment (elfmimi) [reply quantum/matrix.c]: Seeing #7915 I realized that with current master (https://github.com/qmk/qmk_firmware/commit/4eed0331bf0876db0d557f86c7d4ef599ff0e6c4) as of today, this is not being caught.

`#define DIODE_DIRECTION CAL2RAW`

Shall we change COL2ROW to 1 and ROW2COL to 2 and so on?

File: quantum/matrix.c (hunk 2 above)
Comment (elfmimi) [reply quantum/matrix.c]: Working on this in #8115.

File: quantum/config_common.h (hunk 1 above)
Comment (elfmimi) [reply quantum/config_common.h]: Well, I find ...
`#define DIODE_DIRECTION BOTHWAYS`
`#define DIODE_DIRECTION EITHERWAS`
nifty. But it's only a name. I'll change it to what ever we agree on.
That said, ROW2COL2ROW does not convey the working of EITHERWAY. Please come up with a good name.
Anyone?

btw, COL2ROW2COL was baptized by @drashna in https://github.com/qmk/qmk_firmware/pull/6313#issuecomment-511488976 .

**Comment by elfmimi** [comment]
Updated according to the requests. (except the symbol names)
Minimal extent of document is added.

**Comment by skullydazed** [comment]
This is pretty awesome, thank you for bringing it to core. I haven't reviewed the code in detail yet but I'm excited about what it will mean for custom keyboards.

> I'm planning to do documentation later on. after this is accepted.

We'll be able to review the code and test it out, but we'll need documentation on how `EITHERWAY` and `BOTHWAY` work before we can merge this.

**Comment by elfmimi** [comment]
description updated.

**Comment by elfmimi** [comment]
cope with qmk cformat (clang-format).

**Comment by elfmimi** [comment]
For reference, here is an example how to apply BOTHWAYS to the-dog-keyboard.
https://github.com/elfmimi/qmk_firmware/commit/d1b75052e03b459701239a63e5f6d4815f6d61f9

The matrix of this keyboard can be found here.
https://github.com/qmk/qmk_firmware/pull/7401#discussion_r347714947
https://user-images.githubusercontent.com/15257475/69113624-de717000-0ac6-11ea-8f0b-d603c023e05b.png

**Comment by angustrau** [comment]
> The matrix of this keyboard can be found here. [#7401 (comment)](https://github.com/qmk/qmk_firmware/pull/7401#discussion_r347714947)
> https://user-images.githubusercontent.com/15257475/69113624-de717000-0ac6-11ea-8f0b-d603c023e05b.png

Hi! Won't this matrix ghost after you press 3 keys? eg. SW1, SW57 and SW8 would cause SW7 to trigger.

**Comment by elfmimi** [comment]
That's a real good question. The answer is, It won't! Because `Vf * 3 > Vih` .
(for commodity diodes like 1N4148)
@angustrau

**Comment by sowbug** [comment]
> The matrix of this keyboard can be found here. [#7401 (comment)](https://github.com/qmk/qmk_firmware/pull/7401#discussion_r347714947)
> https://user-images.githubusercontent.com/15257475/69113624-de717000-0ac6-11ea-8f0b-d603c023e05b.png

I might be missing something obvious. That link shows a standard matrix, not a duplex one. It has six rows and twenty columns, all the diodes go in the same direction, and every switch belongs to exactly one physical row and column. Did you mean to post a different image?

**Comment by elfmimi** [comment]
@sowbug Please take a look once again closely. In the left half of the matrix diodes are pointing upward (ROW2COL) and in the right half downward (COL2ROW) .

**Comment by sowbug** [comment]
Yes, I see now. I was looking at adjacent diodes expecting them to alternate direction. Thank you!

File: quantum/config_common.h [hunk 4]
```diff
@@ -17,8 +17,10 @@
 #pragma once
 
 /* diode directions */
-#define COL2ROW 0
-#define ROW2COL 1
+#define COL2ROW 1
```
Comment (nooges) [inline quantum/config_common.h:20]: Any particular reasoning for the renumbering of these two defines?

Comment (elfmimi) [reply quantum/config_common.h:20]: Without that, something like this will compile. This as just an example.
#define DIODE_DIRECTION RAW2CAL
And, Instead of warned or reported with an error, it will be treated as COL2ROW.

**Comment by elfmimi** [comment]
@hineybush's episode is a strikingly well suited example which showcase the enhancement this PR enables.
Though it surely is unwanted accident for him, I'm very proud that I could lend him a helping hand in this.

I heard some voices of anticipation hoping this PR get merged.
So I'm here to provide a quick recap.

At first there was no document, but I've already done decent documentation in the later commit.
Thus 'need doc' label does not apply anymore.

I beleave we've had more than enough time for people to swallow and consume the concept of this PR.
Just in case, there may be someone reading this who find oneself feeling kind of not fully understanding. It's okay. It is not because of lack of documentation. It is because of lack of knowledge in electronics. Just let it go.
It really is as simple as "Doing COL2ROW and ROW2COL at the same time." , really.

What left was @fauxpark's statement about his fondness toward the wacky name of COL2ROW2COL.
I think that listing COL2ROW2COL next to COL2ROW and ROW2COL will make those two basic options look more complicated than it is. Especially for someone who is new to QMK. On the other hand, EITHERWAY and BOTHWAYS will definitely look as advanced options. Also you can speculate what they do from their names.
The bottom line is I'm not going to change those names.
I'll welcome usages of COL2ROW2COL as a nickname to BOTHWAYS.

This PR is ready to merge.

**Comment by hineybush** [comment]
> That's a real good question. The answer is, It won't! Because `Vf * 3 > Vih` .
> (for commodity diodes like 1N4148)
> @angustrau

it actually does cause ghosting. I shipped a few of my "broken" hbcp PCBs out and a customer has had an issue with keys triggering while holding down three others (left shift, left ctrl and left arrow triggers right shift).

luckily I can fix this with the "fixed" PCBs with the correct diodes, once I get them in..

**Comment by elfmimi** [comment]
@hineybush Thank you for sharing your experience with us! I really appreciate it.
You are right. from AT90USB128s' datasheet
> 31.2 DC characteristics
> Vih (min)  = 0.6VCC

This will showcase a failure example. Ironically.
Too bad I couldn't help you, @hineybush .

Definitely, I will add a description on this matter.

**Comment by tzarc** [comment]
Has conflicts, and unapplied suggestions.
Removed the on-hold label so stalebot can eventually close it.

**Comment by kbrock** [comment]
This is very cool and the diffs are relatively minor.

Is there any chance for this code?
Does it make sense for someone to try and revive this PR to get it into the code base? Or has this ship sailed?

**Comment by elfmimi** [comment]
@kbrock I believe anyone can pick this up and continue working on it. because it is already openly release code under the same GPL license.

**Comment by elfmimi** [comment]
Hi! I'd say you should better create another pull request for your new revised construct.
As for the commit history, do it any way you feel comfortable. @NoOne2246
